sip/sip-001-burn-election.md

```diff
@@ -0,0 +1,251 @@
+streamed, rather than batched, transactions can be immediately
+"confirmed" by the leader by reading them out of the mempool and
+broadcasting them as part of the current block. To do this, the leader
+wraps the transaction in a "microblock" (essentially just a
+blockheader) and signs the microblock. The design of microblocks is
+motivated by [Bitcoin-NG](https://www.usenix.org/node/194907).
+
+When a leader starts an epoch, the first transaction of the epoch must
+be announced on the burn-chain. This first transaction, which we call
+the `begin-epoch` transaction is a no op, and merely commits the
+leader to a particular chaintip. This prevents a leader from hiding
+their chain selection until a later block. A leader's `begin-epoch`
+transaction does not need to be included on the burn chain before the
+leader's epoch begins, rather the leader is allowed up to _l_ blocks
+(we will use 6) for their transaction to be included on the
+burn-chain. If the leader's `begin-epoch` transaction is not
+included on the burn-chain within that time, the epoch is skipped.
+
+Once the epoch begins, the operation of the leader node is relatively
+straight-forward. The leader starts by reading transactions out of the
+mempool. Transactions should be selected based on the transaction fee
+and expected cost of evaluating that transaction. The leader then
+executes the transaction, and broadcasts the transaction in a signed
+microblock. The microblock contains the hash of the previous
+microblock, the transaction, and a signature from the leader over the
+microblock.
+
+The leader is only able to choose and broadcast microblocks for the
+block (or epoch) in which it was selected as the leader. The node will
+cease to be the leader if the next epoch begins and another leader is
+selected.
+
+## Leader chain selection
+
+As mentioned above, when a leader begins an epoch, it can choose the
+chain tip from which to begin the new block. This allows leaders to
+create forks from older blocks in an attempt to reorganize the Stack
+chain. In the normal operation of the blockchain, we expect many
+short-lived forks to occur at the transition between epochs. As one
+leader begins broadcasting a new epoch, invariably some transactions
+from the prior epoch will be orphaned, and need to be
+rebroadcasted. We call these _micro-forks_.
+
+To choose between forks, Stack nodes employ a chain selection
+algorithm which ignores transactions other than transactions on the
+longest chain. That is discussed in more detail in the `Fork
+Selection` section of this proposal. However, by requiring that
+leaders broadcast at least one microblock (the `begin-epoch`) on the
+burn-chain, it guarantees that subsequent leaders are able to include
+a portion of the epoch in their chain, even if the leader behaves
+maliciously by, for example, hiding some transactions with the
+intention of later broadcasting them to invalidate subsequent leaders'
+chains.
```

Comment:
I think this could be leveraged to realize a hidden benefit:  **it could prevent all but the micro-forks**.  If we make it so a leader must broadcast their coinbase tx on-chain, and we make it so the leader *must* commit to a microblock no earlier than the last coinbase tx, then it's impossible at the protocol level for a leader initiate a fork.  Even if a leader knows they will be selected twice in a row, they would be denied the ability to reorg the Stacks chain -- they could only micro-fork themselves.

Granted, this could cause Stacks blocks to get invalidated if the burnchain is reorganized  -- for example, a 1-confirmation block could get orphaned, and which would erase its coinbase tx and also orphan any coinbase txs that built on it even if they got confirmed.  However, this is easy to recover from, and the leaders still get their Stacks coinbases (they just don't get the transaction fees).

What do you think?